File: preprocessing/feature_engineering.R

library(dplyr)

feature_engineering <- function(df){

  df <- df %>%
    mutate(
      tenure_group = case_when(
        tenure < 12 ~ "new",
        tenure < 24 ~ "mid",
        TRUE ~ "loyal"
      )
    )

  df$tenure_group <- as.factor(df$tenure_group)

  return(df)
}

File: preprocessing/feature_engineering.py
import pandas as pd

def feature_engineering(df):
    df['tenure_group'] = pd.cut(df['tenure'], bins=[-1, 11, 23, float('inf')], labels=['new', 'mid', 'loyal'])
    df['tenure_group'] = df['tenure_group'].astype('category')
    return df
